ambry-frontend/src/main/java/com/github/ambry/frontend/UserQuotaRequestCostPolicy.java

```diff
@@ -0,0 +1,142 @@
+        calculateCapacityUnitCost(restRequest, restResponseChannel, blobInfo));
+    costMap.put(QuotaName.STORAGE_IN_GB.name(), calculateStorageCost(restRequest, restResponseChannel));
+    return costMap;
+  }
+
+  /**
+   * Get appropriate cost metric name for {@link RestMethod}.
+   * @param restMethod {@link RestMethod} object.
+   * @return CUQuotaMethod name for specified {@link RestMethod}.
+   */
+  private String restMethodToCostMetric(RestMethod restMethod) {
+    switch (restMethod) {
+      case GET:
+      case OPTIONS:
+      case HEAD:
+        return QuotaName.READ_CAPACITY_UNIT.name();
+      default:
+        return QuotaName.WRITE_CAPACITY_UNIT.name();
+    }
+  }
+
+  /**
+   * Calculate the storage cost incurred to serve a request.
+   * For post requests this is the number bytes in GB of the blob data. For all other write requests the default is DEFAULT_COST_FOR_HEAD_DELETE_TTL.
+   * For read requests there is no storage cost incurred.
+   * @param restRequest {@link RestRequest} to find type of request.
+   * @param restResponseChannel {@link RestResponseChannel} object.
+   * @return storage cost.
+   */
+  private double calculateStorageCost(RestRequest restRequest, RestResponseChannel restResponseChannel) {
+    switch (restRequest.getRestMethod()) {
+      case POST:
+        long contentSize = 0;
+        if (restRequest.getUri().contains(Operations.STITCH)) {
+          contentSize = Long.parseLong((String) restResponseChannel.getHeader(RestUtils.Headers.BLOB_SIZE));
+        } else {
+          contentSize = restRequest.getBytesReceived();
+        }
+        return contentSize / BYTES_IN_GB;
+      case DELETE:
+      case PUT:
+        return CU_COST_UNIT / BYTES_IN_GB; // Assuming 1 chunk worth of storage cost for deletes and ttl updates.
+      default:
+        return 0;
+    }
+  }
+
+  /**
+   * Calculate the cost incurred in terms of capacity unit to serve a request.
+   * For post and get requests it is the number of CU_COST_UNITs determined by blob size.
+   * For all other requests, the default cost is DEFAULT_COST_FOR_HEAD_DELETE_TTL.
+   * @param restRequest {@link RestRequest} to find type of request.
+   * @param restResponseChannel {@link RestResponseChannel} object.
+   * @param blobInfo {@link BlobInfo} object representing the blob being served.
+   * @return cost in terms of capacity units.
+   */
+  private double calculateCapacityUnitCost(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      BlobInfo blobInfo) {
+    switch (restRequest.getRestMethod()) {
+      case POST:
+        long contentSize = 0;
+        if (restRequest.getUri().contains(Operations.STITCH)) {
+          contentSize = Long.parseLong((String) restResponseChannel.getHeader(RestUtils.Headers.BLOB_SIZE));
+        } else {
+          contentSize = restRequest.getBytesReceived();
+        }
+        double cost = Math.ceil(contentSize / CU_COST_UNIT);
+        return (cost > 0) ? cost : 1;
```

Comment:
I don't quite understand this part, what does the `ceil` do here? When will cost equal to  zero? (contentSize == 0?)